README.rst
=========
coccigrep
=========

Introduction
============

coccigrep is a semantic grep for the C language based on coccinelle
(http://coccinelle.lip6.fr). It can be used to find where a given
structure is used in code files. coccigrep depends on the spatch
program which comes with coccinelle.

Usage
=====

Run `coccigrep -h` for complete options.

Examples
========

To find where in a set of files the type named `Packet` is used, you
can run ::

    $ coccigrep  -t Packet *c
    source-af-packet.c:272:         p = ptv->in_p;
    source-af-packet.c:300:     p->datalink = ptv->datalink;
    source-af-packet.c:758:     switch(p->datalink) {

If you want to match on structure, you need to provide the complete name ::

    coccigrep -t 'struct seq_file' fs/seq_file.c
    fs/seq_file.c:654 (struct seq_file *seq): 	seq = f->private_data;
    fs/seq_file.c:655 (struct seq_file *seq): 	seq->private = private;
    fs/seq_file.c:537 (struct seq_file *m): 	if (m->count < m->size) {


To find where in a set of files the `datalink` attribute is used in the structure
named `Packet`, you can simply do ::

    $ coccigrep  -t Packet -a datalink  *c
    source-af-packet.c:300:     p->datalink = ptv->datalink;
    source-af-packet.c:758:     switch(p->datalink) {
    source-erf-dag.c:525:     p->datalink = LINKTYPE_ETHERNET;

If you want to be more precise and find where this attribute is set, you can use 
the operation flag (-o). One of its value is `set` which indicate we only want
the match where the attribute is set ::

    $ coccigrep  -t Packet -a datalink -o set  source*c
    source-af-packet.c:300:     p->datalink = ptv->datalink;
    source-erf-dag.c:525:     p->datalink = LINKTYPE_ETHERNET;

Installation
============

The dependencies of coccigrep are spatch which comes with coccinelle. On python side, you
need setuptools and optionally pygments (for colorized output). Happy Debian user can do ::

    aptitude install python-setuptools python-pygments

To install coccigrep run ::

    sudo python ./setup.py install

Configuration
=============

As from version 0.8, coccigrep can be configured via a configuration file. A complete sample of
configuration file is available in the src/coccinelle.cfg.

Hierarchical configuration
--------------------------

The configuration file system is hierarchical and the following files are parsed in that order

 - host config in /etc/coccigrep
 - user config in ~/.coccigrep
 - directory config in .coccigrep

Thus, for example, the directory config settings will overwrite host config settings.

Interesting options
-------------------

In the global section, the `concurrency_level` is the most interesting. It codes the number of
spatch commands that will be launched in parallel. If multiple files are search, this will
increase dramatically performances at the cost of a little increase of memory usage.

If you want to add your own semantic patches, you just have to put them in a directory with
name matchting the wanted operation name (`zeroed.cocci` will lead to the `zeroed` operation).
Then add a `local_cocci_dir` pointing to this directory in the global section.

For a description of the writing of semantic patches see `coccigrep homepage`_.

.. _coccigrep homepage: http://home.regit.org/software/coccigrep/

Other options are more explicit and are direct mapping of the associated command line option.

Running coccigrep in vim
------------------------

To use coccigrep in vim, you can use the `cocci-grep.vim` plugin provided in
the `editors` directory. To do so you can simply copy it to your plugin directory
which is usually `~/.vim/plugin/`. If your `coccigrep` script in not in your
path, you can use the coccigrep_path variable to give complete path. For
example, you can add to your `.vimrc` ::

    let g:coccigrep_path = '/usr/local/bin/coccigrep'

And then you can run commands like ::

    :Coccigrep
    :Coccigrep Packet datalink source-*.c
    :Coccigrep Packet datalink set source-*.c

First command will interactively ask you the value. Second one will search all
dereference of the datalink attribute for Packet structure. The last one will
look where the set operation is done on the datalink attribute of Packet. To get
the list of operations on your system, you can run `coccigrep -L` or look at
the list provided when input for operation is asked in interactive mode.

The matches will appear in the `quickfix list` and the file corresponding to first
match will be opened at the corresponding line. Note that you can use completion on
structure and attribute names based on tags (generated by `make tags`).

To run a search in vim on a non-named structure, you must quote the spaces and
thus run something like ::

    :Coccigrep "struct nfq_data"  s*c

Please note that, in interactive mode, quoting is not necessary.

You can also set the global variable `coccigrep_files` ::

    :let g:coccigrep_files = '~/myproject/src/flist'

where `flist` is the file corresponding to the `-l` option.
And then you can run commands like ::

    :Coccigrep Packet
    :Coccigrep Packet datalink set

That is, you don't need to provide the last argument of the previous examples.
This is particularly useful if you set vim's autochdir option.

Running coccigrep in emacs
--------------------------

To use coccigrep in emacs, you need to load the `cocci-grep.el` module provided in the `editors`
directory of the source code. For example, if you copy it in `~/.emacs.d/site-lisp/`, you
can do ::

    (add-to-list 'load-path "~/.emacs.d/site-lisp/")
    (require 'cocci-grep)

And then you can run something like ::

    Meta+x cocci-grep

and answer to the questions which are

 - Type: The structure type you are searching
 - Attribut: The attribute in the structure
 - Operation: The operation on the structure. The set of commands include set,used,func,test,deref
 - Files: A blob expression that will match the file you want to search in

The matches will appear in a buffer with mode set to `grep-mode` and you will thus be able to jump
on occurence. History is available on the different parameters.


ChangeLog
1.12 (2014/01/10)
 * Improve accuracy (Eric Leblond)
 * vim: correctly indent file (Eric Leblond)
 * add match on identifier name (Eric Leblond)
 * Merge pull request #11 from yhuelf/master (Eric Leblond)
 * vim plugin: cosmetics (Frédéric Yhuel)
 * vim plugin: allow use of "-l" option (Frédéric Yhuel)
 * fix version detection (Eric Leblond)
 * --version is not supported by all spatch version (Eric Leblond)
 * Add structure matching example. (Eric Leblond)
 * Add information about dependencies. (Eric Leblond)
 * Reorder README so we see installation part (Eric Leblond)

1.12 (2012/12/22)
 * Fix error handling (Eric Leblond)
 * Put operation in try pass. (Eric Leblond)
 * Don't ignore header files (Frédéric Yhuel)
 * Don't rm temporary .cocci file in verbose mode (Frédéric Yhuel)

1.11 (2012/09/10)
 * regular expression operator has changed (Eric Leblond)
 * func: match with regexp (Eric Leblond)
 * func: now able to match in case of cast (Eric Leblond)
 * Missing file was not missing. (Eric Leblond)
 * Add missing file. (Eric Leblond)
 * Improve function match (Eric Leblond)

1.10 (2012/08/31)
 * vim: add support for search on "struct toto" (Eric Leblond)
 * Fix error message. (Eric Leblond)
 * Factorize version numbering. (Eric Leblond)

1.9 (2012/07/31)
 * Add --defined option. (Eric Leblond)
 * Add undefined option. (Eric Leblond)

1.8 (2012/03/05)
 * Fix CPP support. (Eric Leblond)

1.7 (2012/03/04)
 * Directory support: Move code to script (Eric Leblond)
 * Add support for directory recursion (Cyril Roelandt)
 * Fix a sentence. (Cyril Roelandt)
 * Replace "attribut" by "attribute". (Cyril Roelandt)

1.6 (2011/11/07)
 * Remove matched information from context lines (Guillaume Nault)
 * Fix issue with single concurrency run (Eric Leblond)
 * Prepare 1.6 release. (Eric Leblond)
 * Fix context group separation (Guillaume Nault)
 * Fix docstring typo (Guillaume Nault)
 * Change field separator for context lines (Guillaume Nault)
 * Fix compatibility with python 2.5. (Eric Leblond)
 * Add a separator between blocks of matched lines (Guillaume Nault)
 * Avoid context line duplication (Guillaume Nault)
 * Add missing information in CocciMatch::display() docstring (Guillaume Nault)
 * Fix context line numbers (Guillaume Nault)

1.5 (2011/11/03)
 * Allow specyfing a list of input files via a file (-l / --file-list) (Thomas Graf)
 * Fix unmatched patterns when using option "--operation=set" (Guillaume Nault)

1.4 (2011/10/27)
 * Detect attribute setting with unary operator. (Eric Leblond)
 * Manpage update: describe option "-o, --operation" (Guillaume Nault)
 * Improve manpage (Guillaume Nault)

1.3 (2011/10/06)
 * Add manpage for coccigrep. (Guillaume Nault)

1.2 (2011/09/15)
 * unprint cmd (Phlogistique)

1.1 (2011/09/14)
 * Add C++ support. (Eric Leblond)
 * Add 'Revision' info to cocci patches. (Eric Leblond)
 * Add 'Arguments' list in cocci patches. (Eric Leblond)
 * Add 'Arguments' as a keyword for cocci comment. (Eric Leblond)
 * Add information about -L -v in help message. (Eric Leblond)
 * Fix trivial typo. (Eric Leblond)
 * Use python from env. (Eric Leblond)
 * '-L -v' now display a list of operations description. (Eric Leblond)
 * Prepare print of operations informations. (Eric Leblond)
 * Add class to get information from cocci patch. (Eric Leblond)

1.0 (2011/09/12)
 * Add description of operation in comment lines. (Eric Leblond)

1.0rc2 (2011/09/06)
 * Document vim search of 'struct myst' (Eric Leblond)
 * Fix usage of structure without typedef. (Eric Leblond)
 * '-C' option can now be used to set context. (Eric Leblond)
 * Improve display for pointer/non pointer match (Eric Leblond)
 * Add non pointer support to deref. (Eric Leblond)
 * Rename '-C' option to '-p' (Eric Leblond)
 * big cosmit. please test for your sake (Feth AREZKI)
 * try and simplify config reading code (Feth AREZKI)
 * Set version to 1.0rc2. (Eric Leblond)
 * Handle static structure. (Eric Leblond)
 * Fix local cocci dir. (Eric Leblond)
 * Use setuptools in setup.py. (Eric Leblond)

1.0rc1 (2011/09/02)
 * Set version to 1.0rc1 (Eric Leblond)
 * print debug message to stderr. (Eric Leblond)
 * Update MANIFEST.in to include doc. (Eric Leblond)
 * Improve doc file. (Eric Leblond)
 * Improve documentation files. (Eric Leblond)
 * Add docstrings to module. (Eric Leblond)
 * Add some docstrings. (Eric Leblond)
 * Suppress false space. (Eric Leblond)
 * Prepare doc generation via sphinx. (Eric Leblond)
 * Add some docstrings. (Eric Leblond)
 * Handle a serie of failures via exception. (Eric Leblond)
 * put more code in try except to intercept all exceptions (Eric Leblond)
 * Work exception handling (Eric Leblond)
 * Fix typo in split files function. (Eric Leblond)
 * Update setup.py for upload to python repo. (Eric Leblond)

0.9 (2011/08/29)
 * Improve vim usage instruction. (Eric Leblond)
 * Display message about waiting time. (Eric Leblond)
 * vim plugin: Use script variable. (Eric Leblond)
 * vim plugin display warning message when there is no match. (Eric Leblond)
 * cocci-grep.vim is now a vim plugin. (Eric Leblond)
 * Exit with 1 if no match. (Eric Leblond)
 * Add vim module to package. (Eric Leblond)
 * Add operation list to interactive vim function (Eric Leblond)
 * Add option to list available operations. (Eric Leblond)
 * Add documentation of vim module. (Eric Leblond)
 * Vim function for coccigrep (Eric Leblond)
 * Minor fix in emacs module (Eric Leblond)

0.8 (2011/08/27)
 * Add information about configuration file (Eric Leblond)
 * Add explanation about emacs. (Eric Leblond)
 * Add basic cocci-grep emacs module (Eric Leblond)
 * Add Emacs mode. (Eric Leblond)
 * Add option for spatch command (Eric Leblond)
 * Add sanity check to add_operations (Eric Leblond)
 * Add support for local cocci directory (Eric Leblond)

editors/cocci-grep.el
;; Copyright (C) 2011 Eric Leblond
;;
;; this file is not part of Emacs
;;
;; Author: Eric Leblond
;; Maintainer: Eric Leblond
;; Description: provide interface to coccigrep
;;
;; Based on xargs-grep module by Le Wang
;;
;;; Installation: put this file in your load-path and
;;
;; (require 'cocci-grep)
;;
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;;
;; This program is free software; you can redistribute it and/or
;; modify it under the terms of the GNU General Public License as
;; published by the Free Software Foundation; either version 3, or
;; (at your option) any later version.
;;
;; This program is distributed in the hope that it will be useful,
;; but WITHOUT ANY WARRANTY; without even the implied warranty of
;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
;; General Public License for more details.
;;
;; You should have received a copy of the GNU General Public License
;; along with this program; see the file COPYING.  If not, write to
;; the Free Software Foundation, Inc., 51 Franklin Street, Fifth
;; Floor, Boston, MA 02110-1301, USA.
;;
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;

(require 'compile)
(require 'grep)

(defvar cocci-s-type-history '()
  "The minibuffer history list for `\\[cocci-grep]'s type argument.")

(defvar cocci-s-attribute-history '()
  "The minibuffer history list for `\\[cocci-grep]'s attribute argument.")

(defvar cocci-s-operation-history '()
  "The minibuffer history list for `\\[cocci-grep]'s operation argument.")

(defvar cocci-files-history '()
  "The minibuffer history list for `\\[cocci-grep]'s files argument.")


(defun cocci-grep-read-type ()
  (read-from-minibuffer "Type: "
                        nil nil nil
                        'cocci-s-type-history
                        nil))
(defun cocci-grep-read-attribute ()
  (read-from-minibuffer "Attribut: "
                        nil nil nil
                        'cocci-s-attribute-history
                        nil))
(defun cocci-grep-read-operation ()
  (read-from-minibuffer "Operation: "
                        nil nil nil
                        'cocci-s-operation-history
                        nil))
(defun cocci-grep-read-file-string ()
  (read-from-minibuffer "Files: "
   nil nil nil
   'cocci-files-history
   nil))

;;;###autoload
(defun cocci-grep (s-type s-attribute s-operation files)
  "s-type is the searched type, s-attribute the attribute, s-operation the
operation on structure and files is a blob expression that will match files"
  (interactive (list
                (cocci-grep-read-type)
                (cocci-grep-read-attribute)
                (cocci-grep-read-operation)
                (cocci-grep-read-file-string)))
  (let (out-buf
        )
    (setq out-buf (compilation-start (concat "coccigrep -E -t " s-type " -a " s-attribute
                                             " -o " s-operation " " files) 'grep-mode))
    ))

(provide 'cocci-grep)



src/coccigrep.py
# Copyright (C) 2011,2012 Eric Leblond <eric@regit.org>
#
# You can copy, redistribute or modify this Program under the terms of
# the GNU General Public License version 3 as published by the Free
# Software Foundation.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# version 3 along with this program; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
# 02110-1301, USA.

from ConfigParser import SafeConfigParser
from os import unlink, path, listdir, getcwd
from string import Template
from subprocess import Popen, PIPE, STDOUT
from sys import stderr
from tempfile import NamedTemporaryFile
import errno
import re

COCCIGREP_VERSION = "1.13"

have_multiprocessing = True
try:
    from multiprocessing import Process, Pipe
except ImportError:
    have_multiprocessing = False


have_pygments = True
try:
    from pygments import highlight
    from pygments.lexers import CLexer
    from pygments.filters import NameHighlightFilter
    from pygments.formatters import Terminal256Formatter, HtmlFormatter
except ImportError:
    have_pygments = False


class CocciException(Exception):
    """
    Generic class for coccigrep exception
    """
    def __init__(self, value):
        self.value = value

    def __str__(self):
        return self.value


class CocciConfigException(CocciException):
    """
    Exception raised when configuration parameter are not correct.

    For example, it is returned if spatch command can not be found.
    """
    pass


class CocciRunException(CocciException):
    """
    Exception raised when running parameters are not correct.

    For example, it is returned if a required argument is missing.
    """
    pass


# produces conf file paths in reverse specificity order, ie
# - package conf
# - system wide
# - user
# - current dir
_CONF_FILES = tuple((path.join(dirname, name_format % 'coccigrep')
    for dirname, name_format in
    ((path.dirname(__file__), '%s.cfg'),
    ('/etc', '%s'),
    (path.expanduser('~'), '.%s'),
    (getcwd(), '.%s'))))


class CocciGrepConfig(SafeConfigParser):
    """
    Configuration handling class

    This class parses configuration and can be used to access to
    configuration item via get operations. CocciGrepConfig is derived
    from SafeConfigParser
    """
    def __init__(self):
        SafeConfigParser.__init__(self)
        self._load_config()

    def _load_config(self):
        """
        loads configuration from files given in _conf_FILES,
        overwriting the less specific by the most specific (given last)
        """
        for index, filename in enumerate(_CONF_FILES):
            if path.isfile(filename):
                self.read(filename)
            elif index == 0:
                raise CocciException('No package config file: %s' % (filename))


class CocciMatch:
    """
    Store a match and take care of its display
    """

    ptype_regexp = re.compile("^[ )]*\.")

    def __init__(self, mfile, mline, mcol, mlineend, mcolend):
        self.file = mfile
        self.line = int(mline)
        self.column = int(mcol)
        self.lineend = int(mlineend)
        self.columnend = int(mcolend)
        self.start_at = self.line
        self.stop_at = self.line
        self.trailer = ""

    def display(self, stype, mode='raw', oformat='term'):
        """
        Display output for a single match

        :param stype: name of the matched type
        :type stype: str
        :param mode: display mode
        :type mode: str
        :param oformat: format of output for color (term, html)
        :type oformat: str
        :return: a human readable string containing the result of the search
                 (matched line, context, file name, etc.)
        """
        f = open(self.file, 'r')
        lines = f.readlines()
        pmatch = lines[self.line - 1][self.column:self.columnend]
        ptype = "*" # match is a pointer to struct
        if (CocciMatch.ptype_regexp.search(lines[self.line - 1][self.columnend:])):
            ptype = ""
        output = ""
        if mode == 'color':
            output += "%s: l.%s -%d, l.%s +%d, %s %s%s\n" % (self.file,
                 self.line, self.line - self.start_at, self.line,
                 self.stop_at - self.line, stype, ptype, pmatch)
        for i in range(self.start_at - 1, min(self.stop_at, len(lines))):
            if mode == 'color':
                output += lines[i]
            elif mode == 'vim':
                output += "%s|%s| (%s %s%s): %s" % (self.file, i + 1,
                stype, ptype, pmatch, lines[i])
            elif mode == 'emacs':
                output += "%s:%s: (%s %s%s): %s" % (self.file, i + 1,
                stype, ptype, pmatch, lines[i])
            elif i == self.line - 1:
                output += "%s:%s (%s %s%s): %s" % (self.file, i + 1,
                stype, ptype, pmatch, lines[i])
            else:
                output += "%s-%s %s - %s" % (self.file, i + 1,
                ' ' * (2 + len(stype + ptype + pmatch)), lines[i])
        f.close()
        if mode == 'color':
            if have_pygments:
                lexer = CLexer()
                lfilter = NameHighlightFilter(names=[pmatch])
                lexer.add_filter(lfilter)
                if oformat == "term":
                    return highlight(output, lexer, Terminal256Formatter())
                elif oformat == "html":
                    return highlight(output, lexer,
                        HtmlFormatter(noclasses=True))
                else:
                    return output
        return output + self.trailer


class CocciProcess:
    """
    Class used for running spatch command in the case of multiprocessing
    """
    def __init__(self, cmd, verbose):
        self.process = Process(target=self.execute, args=(self, ))
        self.output, self.input = Pipe()
        self.cmd = cmd
        self.verbose = verbose

    def execute(self, option=''):
        output = ""
        try:
            if self.verbose:
                stderr.write("Running: %s.\n" % " ".join(self.cmd))
                output = Popen(self.cmd, stdout=PIPE).communicate()[0]
            else:
                output = Popen(self.cmd, stdout=PIPE,
                    stderr=PIPE).communicate()[0]
        except Exception, err:
            import pickle
            output = pickle.dumps(err)

        self.input.send(output)
        self.input.close()

    def start(self):
        self.process.start()

    def join(self):
        self.process.join()

    def recv(self):
        return self.output.recv()


def _operation_name(fname):
    return path.split(fname)[-1].replace('.cocci', '')


def _raise_run_err(err, cmd):
    if err.errno in (errno.ENOENT, errno.ENOEXEC):
        raise CocciConfigException("Unable to run spatch command "
            "'%s': %s." % (cmd[0], err.strerror))
    raise CocciRunException("Unable to run '%s': %s." % (" ".join(cmd),
        err.strerror))

class CocciPatch:
    """
    Class used to store information about a patch.

    This class is iterable and can be used as a dictionnary.
    """

    keywords = ["Name", "Author", "Desc", "Confidence", "File", "Revision", "Arguments"]
    comment = re.compile("^ *// *(%s): (.*)" % ("|".join(keywords)))

    def __init__(self, filename):
        # open file
        self.__dict__["File"] = filename
        self.__dict__["Name"] = _operation_name(filename)
        # read file
        f = open(filename, 'r')
        lines = f.readlines()
        # match
        for line in lines:
            mm = CocciPatch.comment.match(line)
            if mm:
                self.__dict__[mm.group(1)] = mm.group(2)

    def __iter__(self):
        """ return iterator over keys """
        return self.__dict__.__iter__()

    def __getitem__(self, key):
        # name ok?
        if not key in CocciPatch.keywords:
            raise KeyError("Trying to get invalid name '%s'." % key)
        return self.__dict__[key]

    def __setitem__(self, key, value):
        # name ok?
        if not key in CocciPatch.keywords:
            raise KeyError("Trying to set invalid name '%s'." % key)
        # set
        self.__dict__[key] = value

    def __str__(self):
        out = self["Name"] + ": "
        try:
            out += self["Desc"] + "\n"
        except KeyError:
            return out + ": No info available\n"
        for key in self:
            if key not in ["Name", "Desc", "File"]:
                out +=" * %s: %s\n" % (key, self[key])
        return out

class CocciGrep:
    """
    Core class of the module: setup and run.

    This class is the core of the module. It is responsible
    of initialisation and running of the request.
    """
    spatch = "spatch"
    cocci_python = """

@ script:python @
p1 << init.p1;
@@

for p in p1:
    print "%s:%s:%s:%s:%s" % (p.file,p.line,p.column,p.line_end,p.column_end)
"""

    def __init__(self):
        self.verbose = False
        self.spatch = CocciGrep.spatch
        self.ncpus = 1
        self.operations = {}
        self.process = []
        self.matches = []
        self.options = ["--recursive-includes"]
        dirList = listdir(self.get_datadir())
        for fname in dirList:
            op = _operation_name(fname)
            self.operations[op] = path.join(self.get_datadir(), fname)

    def setup(self, stype, attribute, operation):
        """
        :param stype: structure name, used to replace '$type' in the cocci file
        :type stype: str
        :param attribute: basically attribute of the structure, used to replace
            '$attribute' in the cocci file
        :type attribute: str
        :param operation: search operation to do
        :type operation: str
        :raise: :class:`CocciRunException`
        """
        if stype == None:
            raise CocciRunException("Can't use coccigrep without type to "
                "search")
        self.type = stype
        self.attribute = attribute
        self.operation = operation

    def set_concurrency(self, ncpus):
        """
        Set concurrency level (number of spatch command to run in parallel)

        :param ncpus: number of process to launch in parallel
        :type ncpus: int
        """
        if have_multiprocessing:
            self.ncpus = ncpus
            return True
        return False

    def add_options(self, olist):
        """
        Add option to spatch command

        :param olist: List of options
        :type olist: list of str
        """
        self.options += olist

    def set_cpp(self):
        """
        Activate coccinelle C++ support
        """
        self.add_options(["-c++"])

    def set_spatch_cmd(self, cmd):
        """
        Set path or command name for spatch

        :param cmd: Name of parth of the spatch command
        :type cmd: str
        """
        self.spatch = cmd

    def get_datadir(self):
        this_dir, this_filename = path.split(__file__)
        datadir = path.join(this_dir, "data")
        return datadir

    def get_operations(self):
        """
        Get list of available operations

        :return: list of operations in a list of str
        """
        return self.operations.keys()

    def get_operation_name(self, fname):
        return _operation_name(fname)

    def get_operation_info(self, op):
        return CocciPatch(self.operations[op])

    def add_operations(self, new_ops):
        """
        Add operation to the list of supported operations

        :param new_ops: list of filenames (ending by .cocci)
        :type new_ops: list of str
        """
        if len(new_ops) == 0:
            return
        file_filter = re.compile('^[^\.].*\.cocci$')
        for fname in new_ops:
            # file need to end with cocci
            if file_filter.match(fname):
                op = _operation_name(fname)
                self.operations[op] = fname

    def set_verbose(self):
        """
        Activate verbose mode
        """
        self.verbose = True

    def get_spatch_version(self):
        cmd = [self.spatch] + [ '-version']
        try:
            output = Popen(cmd, stdout=PIPE, stderr=STDOUT).communicate()[0]
        except OSError, err:
            _raise_run_err(err, cmd)
        reg = r"version (.*?) with"
        m = re.search(reg, output)
        return m.group(1)

    def spatch_newer_than(self, version):
        from distutils.version import LooseVersion
        sversion = self.get_spatch_version()
        return LooseVersion(sversion) > LooseVersion(version)

    def run(self, files):
        """
        Run the search against the files and directories given in argument

        This function is doing the main job. It will run spatch with
        the correct parameters by using subprocess or it will use
        multiprocessing if a concurrency level greater than 1 has been
        asked.

        :param args: list of filenames and directory names
        :type args: list of str
        :raise: :class:`CocciRunException` or :class:`CocciConfigException`
        """

        if len(files) == 0:
            raise CocciRunException("Can't use coccigrep without files "
                "to search")

        # get version of spatch
        if self.spatch_newer_than("1.0.0-rc6"):
            cocci_op = "=~"
        else:
            cocci_op = "~="
        # create tmp cocci file:
        tmp_cocci_file = NamedTemporaryFile(suffix=".cocci", delete=not self.verbose)
        tmp_cocci_file_name = tmp_cocci_file.name
        # open file with name matching operation
        cocci_file = open(self.operations[self.operation], 'r')
        # get the string and build template
        cocci_tmpl = cocci_file.read()
        cocci_smpl_tmpl = Template(cocci_tmpl)
        cocci_file.close()
        # do substitution
        cocci_smpl = cocci_smpl_tmpl.substitute(type=self.type,
            attribute=self.attribute, cocci_regexp_equal=cocci_op)
        cocci_grep = cocci_smpl + CocciGrep.cocci_python

        tmp_cocci_file.write(cocci_grep)
        tmp_cocci_file.flush()

        # launch spatch
        output = ""
        # Launch parallel spatch
        if self.ncpus > 1 and len(files) > 1:
            fseq = []
            splitsize = 1.0 / self.ncpus * len(files)
            for i in range(self.ncpus):
                start = int(round(i * splitsize))
                end = int(round((i + 1) * splitsize))
                rfiles = files[start:end]
                if len(rfiles) >= 1:
                    fseq.append(rfiles)
            for sub_files in fseq:
                cmd = [self.spatch]
                cmd += self.options
                cmd += ["-sp_file", tmp_cocci_file.name]
                cmd += sub_files
                sprocess = CocciProcess(cmd, self.verbose)
                sprocess.start()
                self.process.append(sprocess)
            for process in self.process:
                ret = process.recv()
                process.join()
                if not ret.startswith('cexceptions\n'):
                    # CocciProcess return a serialized exception
                    # in case of exception
                    output += ret
                    continue
                import pickle
                err = pickle.loads(ret)
                _raise_run_err(err, cmd)
            tmp_cocci_file.close()
        # Fallback to one spatch
        else:
            cmd = [self.spatch]
            cmd += self.options
            cmd += ["-sp_file", tmp_cocci_file.name]
            cmd += files
            try:
                if self.verbose:
                    stderr.write("Running: %s.\n" % " ".join(cmd))
                    output = Popen(cmd, stdout=PIPE).communicate()[0]
                else:
                    output = Popen(cmd, stdout=PIPE, stderr=PIPE).communicate()[0]
            except OSError, err:
                unlink(tmp_cocci_file_name)
                _raise_run_err(err, cmd)

            tmp_cocci_file.close()

        prevfile = None
        prevline = None
        self.matches = []
        for ematch in output.split("\n"):
            try:
                (efile, eline, ecol, elinend, ecolend) = ematch.split(":")
                nmatch = CocciMatch(efile, eline, ecol, elinend, ecolend)
                # if there is equality then we will already display the line
                if (efile == prevfile) and (eline == prevline):
                    continue
                prevfile = efile
                prevline = eline
                self.matches.append(nmatch)
            except ValueError:
                pass

    def display(self, mode='raw', before=0, after=0, oformat='term'):
        """
        Display output for complete request

        :param mode: display mode
        :type mode: str
        :param before: number of lines to display before match
        :type before: int
        :param after: number of lines to display after match
        :type after: int
        :param oformat: format of output for color (term, html)
        :type oformat: str
        :return: the result of the search as a str
        """
        if before != 0 or after != 0:
            prev_match = None
            for index in xrange(len(self.matches)):
                cur_match = self.matches[index]
                cur_match.start_at = cur_match.line - before
                cur_match.stop_at = cur_match.line + after
                if cur_match.start_at < 1:
                    cur_match.start_at = 1

                if prev_match is not None:
                    prev_match.trailer = "--\n"
                    if prev_match.file == cur_match.file:
                        if prev_match.stop_at >= cur_match.line:
                            prev_match.stop_at = cur_match.line - 1
                        if prev_match.stop_at >= cur_match.start_at:
                            cur_match.start_at = prev_match.stop_at + 1

                        if prev_match.stop_at + 1 == cur_match.start_at:
                            # No seprator if groups are contiguous
                            prev_match.trailer = ""

                prev_match = cur_match

        output = ''.join(match.display(self.type, mode=mode, oformat=oformat)
            for match in self.matches)

        return output.rstrip()
